cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -109,13 +112,100 @@ func (spm *KubeStaticPodPathManager) BackupManifestDir() string {
+	if err := waiter.WaitForStaticPodControlPlaneHashChange(cfg.NodeName, component, beforePodHash); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	// Wait for the static pod component to come up and register itself as a mirror pod
+	if err := waiter.WaitForPodsWithLabel("component=" + component); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	return nil
+}
 
-	// This string-string map stores the component name and backup filepath (if a rollback is needed).
-	// If a rollback is needed,
+func performEtcdStaticPodUpgrade(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, recoverManifests map[string]string) error {
+	// Add etcd static pod spec only if external etcd is not configured
+	if len(cfg.Etcd.Endpoints) != 0 {
+		return fmt.Errorf("external etcd cannot be upgraded with kubeadm")
+	}
+
+	// Need to check currently used version and version from constants, if differs then uprgade
+	runningEtcdVersion, err := util.GetEtcdRunningVersion()
+	if err != nil {
+		return err
+	}
+	targetKubernetesVersion := version.MustParseGeneric(cfg.KubernetesVersion)
+	upgradeEtcdVersion, err := constants.GetEtcdSupportedVersion(uint8(targetKubernetesVersion.Minor()))
+	if err != nil {
+		return err
+	}
+	newEtcdVersion := version.MustParseGeneric(upgradeEtcdVersion)
+	if err != nil {
+		return err
+	}
+	// Ignore error here because the version struct was built with MustParse method
+	// which means any errors will trigger panic.
+	if res, _ := newEtcdVersion.Compare(runningEtcdVersion); res != 1 {
+		return fmt.Errorf("etcd upgrade was requested for the same or more recent version")
+	}
+
+	beforeEtcdPodHash, err := waiter.WaitForStaticPodSingleHash(cfg.NodeName, "etcd")
+	if err != nil {
+		return err
+	}
+
+	// Write the updated etcd static Pod manifest into the temporary directory
+	if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(pathMgr.TempManifestDir(), cfg); err != nil {
+		return fmt.Errorf("error creating local etcd static pod manifest file: %v", err)
+	}
+
+	// Perform etcd upgrade using common to all control plane components function
+	return performStaticPodUpgrade("etcd", waiter, pathMgr, cfg, beforeEtcdPodHash, recoverManifests)
+}
+
+// StaticPodControlPlane upgrades a static pod-hosted control plane
+func StaticPodControlPlane(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, etcdUpgrade bool) error {
 	recoverManifests := map[string]string{}
 
+	// etcd upgrade is done prior to other control plane components
+	if etcdUpgrade {
+		// Perform etcd upgrade using common to all control plane components function
+		if err := performEtcdStaticPodUpgrade(waiter, pathMgr, cfg, recoverManifests); err != nil {
```

Comment:
I think we should backup the etcd data dir before we touch the manifests.

If the etcd upgrade goes wrong:

1. We should replace the current data dir (if any) with restore old one
2. Restore manifests
3. Check the restored etcd works. If it's not responding, tell the user something like "Hey something went wrong. We tried restoring but that failed. Here is the location to your backed up data"